compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -1523,6 +1531,162 @@ toCtor env con =
+                        ERecCon
+                            anyTemplateTy
+                            [ ( anyTemplateField
+                              , EToAny
+                                    (templateDataType tplTypeCon)
+                                    (EVar tplParam))
+                            ]
+                "toAnyTemplate"
+                    | otherwise ->
+                        EBuiltin BEError `ETyApp` lfType `ETmApp`
+                        EBuiltin
+                            (BEText
+                                 "toAnyTemplate is not supported in this DAML-LF version")
+                "fromAnyTemplate"
+                    | envLfVersion env `supports` featureAnyType ->
+                        ETmLam (anyTpl, typeConAppToType anyTemplateTy) $
+                        ECase
+                            (EFromAny
+                                 (templateDataType tplTypeCon)
+                                 (ERecProj
+                                      anyTemplateTy
+                                      anyTemplateField
+                                      (EVar anyTpl)))
+                            [ CaseAlternative CPNone $
+                              ENone $ templateDataType tplTypeCon
+                            , CaseAlternative (CPSome tplParam) $
+                              ESome (templateDataType tplTypeCon) $
+                              EVar tplParam
+                            ]
+                    | otherwise ->
+                        EBuiltin BEError `ETyApp` lfType `ETmApp`
+                        EBuiltin
+                            (BEText
+                                 "fromAnyTemplate is not supported in this DAML-LF version")
+                "_templateTypeRep"
+                    | envLfVersion env `supports` featureTypeRep ->
+                        let resType =
+                                TypeConApp
+                                    (Qualified
+                                         stdlibRef
+                                         (mkModName ["DA", "Internal", "LF"])
+                                         (mkTypeCon ["TemplateTypeRep"]))
+                                    []
+                            resField = mkField "getTemplateTypeRep"
+                         in ERecCon
+                                resType
+                                [ ( resField
+                                  , ETypeRep $ templateDataType tplTypeCon)
+                                ]
+                    | otherwise ->
+                        EBuiltin BEError `ETyApp` lfType `ETmApp`
+                        EBuiltin
+                            (BEText
+                                 "templateTypeRep is not supported in this DAML-LF version")
+                other -> error "convertExternal: Unknown external method"
+    | otherwise = error $ "convertExternal: Unable to inline call to external method: " <> primId
```

Comment:
This error message could probably be a bit more clear. It looks like there are two cases where we will hit this:
1. The argument passed to `external` cannot be split into the right number of components.
2. The template does not exist.

So we should probably point out more clearly what exactly went wrong.

Reply:
agreed, separated the two cases.